shared/js/background/components/resource-loader.js

```diff
@@ -0,0 +1,198 @@
+import Dexie from 'dexie'
+import bundledEtags from '../../../data/etags.json'
+import { alarms } from 'webextension-polyfill'
+/**
+ * @typedef {'tds' | 'surrogates' | 'config'} ResourceName
+ *
+ * @typedef {object} ResourceConfig
+ * @property {ResourceName} name
+ * @property {string} [remoteUrl]
+ * @property {string} [localUrl]
+ * @property {number} [updateIntervalMinutes]
+ * @property {'json'|'text'} [format]
+ *
+ * @typedef {(resourceName: ResourceName, etag: string, value: any) => void} OnUpdatedCallback
+ * @typedef {import('../settings.js')} Settings
+ */
+
+export default class ResourceLoader {
+    /** @type {Dexie?} */
+    static dbc = null
+
+    /**
+     * @param {ResourceConfig} config
+     * @param {{
+     *  settings: Settings
+     * }} opts
+     */
+    constructor (config, { settings }) {
+        this.settings = settings
+        this.name = config.name
+        this.remoteUrl = config.remoteUrl
+        this.localUrl = config.localUrl
+        this.updateIntervalMinutes = config.updateIntervalMinutes || 0
+        this.format = config.format || 'json'
+        this.data = null
+        /** @type {OnUpdatedCallback[]} */
+        this._onUpdatedListeners = []
+
+        if (!this.remoteUrl && !this.localUrl) {
+            throw new Error('invalid config: need a remote or local URL (or both)')
+        }
+
+        this.ready = this.checkForUpdates()
+
+        if (this.updateIntervalMinutes) {
+            const alarmName = `updateResource: ${this.name}`
+            alarms.create(alarmName, { periodInMinutes: this.updateIntervalMinutes })
```

Comment:
I think creating the alarm without first checking it doesn't exist might cause an exception for MV3? (Have you tested this on MV3?)

Reply:
(The old code you're replacing uses the `browserWrapper.createAlarm` wrapper which takes care of that. Perhaps use that here?

Reply:
Good shout - I forgot about this edge case. I didn't see any exceptions in MV3 (seems like this has been fixed now in Chrome), but re-registering does push the time of the next alarm back, so this change is important to ensure those alarms actually run! Fixed in https://github.com/duckduckgo/duckduckgo-privacy-extension/pull/2309/commits/38c199def268eae31127c7dde81adc49a4b51c8d